arcs/src/systems/spatial_relation.rs

```diff
@@ -0,0 +1,204 @@
+            &self.accessor(),
+            world,
+        );
+
+        let drawing_storage = world.read_storage::<DrawingObject>();
+        let mut space = world.write_resource::<Space>();
+
+        space.clear();
+
+        for entity in world.entities().join() {
+            if let Some(drawing) = drawing_storage.get(entity) {
+                space.insert(SpatialEntity::new(drawing.geometry.bounding_box(), entity));
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        primitives::{Line},
+        components::{register, Layer, Name, DrawingObject, Geometry, LineStyle, Dimension, Space},
+        Vector,
+        systems::SpatialRelation,
+    };
+    use specs::prelude::*;
+    use piet::Color;
+
+    #[test]
+    fn setup_creates_all_outstanding_spatial_entities() {
+        let mut world = World::new();
+
+        // make sure we register all components
+        register(&mut world);
+    
+        let layer = Layer::create(
+            world.create_entity(),
+            Name::new("default"),
+            Layer {
+                z_level: 0,
+                visible: true,
+            },
+        );
+    
+        // Add a line to our world
+        let line = Line::new(Vector::new(2.0, 1.0), Vector::new(5.0, -1.0));
+        let _first = world
+            .create_entity()
+            .with(DrawingObject {
+                geometry: Geometry::Line(line),
+                layer,
+            })
+            .with(LineStyle {
+                width: Dimension::DrawingUnits(5.0),
+                stroke: Color::rgb8(0xff, 0, 0),
+            })
+            .build()
+        ;
+
+        // Setup our spatial system
+        let mut system = SpatialRelation::new(&world);
+        System::setup(&mut system, &mut world);
+
+        let space = world.read_resource::<Space>();
+        assert_eq!(space.len(), 1);
+
+    }
+
+    #[test]
+    fn run_will_keep_spatial_updated() {
+        let mut world = World::new();
+
+        // make sure we register all components
+        register(&mut world);
+    
+        let layer = Layer::create(
+            world.create_entity(),
+            Name::new("default"),
+            Layer {
+                z_level: 0,
+                visible: true,
+            },
+        );
+    
+        // Add a line to our world
+        let line = Line::new(Vector::new(2.0, 1.0), Vector::new(5.0, -1.0));
+        let _first = world
+            .create_entity()
+            .with(DrawingObject {
+                geometry: Geometry::Line(line),
+                layer,
+            })
+            .with(LineStyle {
+                width: Dimension::DrawingUnits(5.0),
+                stroke: Color::rgb8(0xff, 0, 0),
+            })
+            .build()
+        ;
+
+        // Setup our spatial system
+        let mut system = SpatialRelation::new(&world);
+        System::setup(&mut system, &mut world);
+        
+        // make some changes after the initial setup
```

Comment:
As well as object creation, we may want to make sure object modification and deletion are detected too.

Reply:
I'm a bit stuck on that part @Michael-F-Bryan ,

Similar as to the test for `name_table_bookkeeping` after we read a *resource* from *world* we can no longer borrow *world* as mutable.

```rust
let space = world.read_resource::<Space>(); // <- world is immutably borrowed here
world.delete_entity(first).unwrap(); // <- We would mutably borrow here
```

I can't seem to find a nice pattern/api for either *dropping* space or another way of letting us do queries but also still being able to manipulate the *world* afterwards.

How would you implement/test this for `NameTable` the implementation for `Space` should be similar

Reply:
Hmm... I guess you could fetch the `space` every time you need it instead of putting it in a local variable.

We should only need the `space` after the second `system.run_now()` when we do our assertions though, shouldn't we?

Reply:
Yes, never storing `Space` in a variable solves this limitation.